[Soft-Float] - Initial Interpreter Implementation of Ps2's floating point unit specification

GitHubProUser67 opened the pull request:
This Pull Request implements the first take ever on real Soft-Float support in PCSX2.

This work is a combination or several efforts and researches done prior.

Credits:

- https://www.gregorygaines.com/blog/emulating-ps2-floating-point-nums-ieee-754-diffs-part-1/

- https://github.com/assumenothing/Tommunism.SoftFloat/tree/main

- https://github.com/GitHubProUser67/MultiServer3/blob/main/BackendServices/CastleLibrary/EmotionEngine.Emulator/Ps2Float.cs

- https://github.com/Goatman13/pcsx2/tree/accurate_int_add_sub

- PCSX2 Team for their help and support in this massive journey.

This pull request should be tested with every games requiring a clamping/rounding mode/float patches (cf: GameDatabase).

Currently, this PR fixes on the interpreters:

- https://github.com/PCSX2/pcsx2/issues/5169

- https://github.com/PCSX2/pcsx2/issues/354

- https://github.com/PCSX2/pcsx2/issues/11507

- https://github.com/PCSX2/pcsx2/issues/10519

- https://github.com/PCSX2/pcsx2/issues/8068

- https://github.com/PCSX2/pcsx2/issues/7642

- https://github.com/PCSX2/pcsx2/issues/5257

- https://github.com/PCSX2/pcsx2/issues/8228

- https://github.com/PCSX2/pcsx2/issues/8595

- https://github.com/PCSX2/pcsx2/issues/10885

- https://github.com/PCSX2/pcsx2/issues/11714

- https://github.com/PCSX2/pcsx2/issues/5070

- https://github.com/PCSX2/pcsx2/issues/4546

- https://github.com/PCSX2/pcsx2/issues/2245

- https://github.com/PCSX2/pcsx2/issues/11528

- https://github.com/PCSX2/pcsx2/issues/11507

- https://github.com/PCSX2/pcsx2/issues/4760

- https://github.com/PCSX2/pcsx2/issues/9831

- Gran Turismo 4 license tests

- Mortal Combat Shaolin Monks

- Tokyo Xtreme Racer Zero (Accurate Mul/Div for the EE FPU)

Any other games using the NegDiv hack or any other FPU rounding mode.

This sets the floor for Soft-Float in PCSX2, a long awaited contribution.

seta-san commented:
Does this work on the recompilers or just interpreter?

refractionpcsx2 commented:
You should try reading the title.

seta-san commented:
> You should try reading the title.

I don’t know why my brain just skimmed over that.

MrCK1 commented:
I ran a bunch of tests for "Test Drive Unlimited" AI during the demo scene after sitting idle on the menu. No combination of settings/interpreters seems to have any effect on behavior. There must be something else going on

AmyRoxwell commented:
I can confirm with the multiplication setting that MHG and Dos not longer have bugged bounces!
![Screenshot_20241112_234106](https://github.com/user-attachments/assets/ee89dd3f-289b-4b9e-8953-f8adf9b629ca)
![Monster Hunter G_SLPM-65869_20241112234029](https://github.com/user-attachments/assets/b3790480-c496-4792-abd4-dd60584c3cce)
![Monster Hunter 2_SLPM-66280_20241112235154](https://github.com/user-attachments/assets/9f2ae3c7-9319-419c-ab7e-fa5bd21d144a)

Shoegzer commented:
Shouldn't this help with #2990 as well?

AmyRoxwell commented:
> Shouldn't this help with #2990 as well?

I remember seeing on the public dev channel that stuntman not longer had AI issues with this and not longer the car AI failed? it's been a while

Blackbird88 commented:
Tourist Trophy NTSC works now too! For the first time License Tests in it work!
![image](https://github.com/user-attachments/assets/7aef3a55-6613-4ed2-883a-ae1940a30cd4)

Tokman5 commented:
I tested the demo replays of Tokyo Xtreme Racer Zero (see issue #5597 ) and noticed that the car movement in interpreter mode is now closer to the movement in recompiler mode.
However, there are still slight differences between the two, and both are far from matching the console playback.

Goatman13 commented:
Nice work!

I tested only 2 games for now. Hype Time Quest which if i remember correctly needs accurate fpu mul, and it works fine without patch. 

Second game is THPS4. While here accurate VU1 add/sub fix issue which is normally fixed by VU1 rounding in database, accurate mul breaks game graphics. Of course, game don't need accurate VU1 mul to work correctly, but something seems to be off when enabled.

Accurate vu1 add/sub/sqrt:
![image](https://github.com/user-attachments/assets/46efa87b-9236-48fd-99e0-7aa8f159ac44)
Accurate vu1 mul (same place in game):
![image](https://github.com/user-attachments/assets/205cb735-49be-4d1e-909d-a2308861ff4c)

GitHubProUser67 commented:
The Tony Hawk case is fixed, the game uses an un-doccumented behaviour in it's 3D engine.

The PS2 has no denormals support .... except in the Mul unit apparently.

The behaviour is now emulated properly.

Goatman13 commented:
https://github.com/PCSX2/pcsx2/issues/3200 is fixed when tested with https://github.com/PCSX2/pcsx2/pull/12001/commits/b7f38061df28c3c620649de2056ecf9c615ed778 
Require accurate mul/div for FPU.
I didn't tested, but Krome studio games should profit from this pr too, ones that have patches in game db. Like Spyro, Star Wars, one of Transformers game.

Farran666 commented:
This pr's EE interpreter fixes #11636 's gamedb issue.

AmyRoxwell commented:
> > I remember seeing on the public dev channel that stuntman not longer had AI issues with this and not longer the car AI failed? it's been a while
> 
> @AmyRoxwell Can you provide a reference to this? There's no indication of it being fixed in #2990, and I assume the devs would have closed it if it were. In any event it involves pathing in Driver 3 as well.

I meant like, while using this PR, not that is has been fixed. Sorry if it was misunderstood. But if it's not mention on the PR maybe the thing it needs it's not here by this initial implementation.

GitHubProUser67 commented:
Driv3r seemed fine when it was tested, Stuntman NTSC is a lot better but still can "slightly" deviate.

I suspect it is once again, the interpreter rounding/clamping values somewhere.

Shoegzer commented:
@AmyRoxwell Ah, I understand you now.  That's great news.

@GitHubProUser67 Thanks, nice to see Driv3r is looking better.  Would it make sense to list these games in your OP?

ArtificialSufficiency commented:
Nice to hear of stuntman finally faring better, unless it was the PAL version that was tested, and not the NTSC?

Goatman13 commented:
Game Constantine, tested US version. On level 2 there is dumpster that we need to climb to progress in game. Recompiler get it fine with positive rounding for EE. This is currently not working with soft floats, no matter what.

Reproduction steps.
* Launch first level of game. 
* Use this cheat to skip to level 2: ```Press select, go to the Journal Screen, and enter RIGHT, LEFT, RIGHT, LEFT, RIGHT, LEFT, SQUARE, CIRCLE, CIRCLE, SQUARE, SQUARE, SQUARE, SQUARE, CIRCLE, CIRCLE.```
* Jump thru fence, kill rats, there will be now dumpster on the right side, one that do not work.

GitHubProUser67 commented:
I managed to fix it, but the game requires accurate soft floats on the VU0 and the EE FPU, so far I tried with booth Add/Sub and Mul/Div on EE FPU/COP2 and VU0:

![image](https://github.com/user-attachments/assets/dd9a6894-0a7b-4438-af65-0505155d1861)

GitHubProUser67 commented:
Note to future tests:

The constantine case is a typical case of a hack fixing a problem that is related to what I call "float broadcasting".

A game might want to transfer a out of IEEE range float from the FPU to the VUs (and vice-versa).

When a game Database entry has a rounding mode, usually it covers a can of worms where the game requires accurate soft floats on more than one processor.

Goatman13 commented:
That's weird. I can't make it work even with both, accurate FPU and accurate VU0, actually that's how i tested it earlier.

GitHubProUser67 commented:
You made sure to be on the last revision of the PR? I fixed some functions since.

Goatman13 commented:
> Edit: You also need to re-load the level, the bad calculation happens at loading time.

Ok, that was it. I didn't expected it require reloading because changing rounding modes worked without that. My bad.

GitHubProUser67 commented:
It's nothing ^^ , diagnosing floats is more tricky than some usual tests, because the wrong calculation can happen way before we even noticed it went wrong.

In the case of constantine, they really seems to pre-calculate a lot of stuff while loading, I guess to save on compute time while ingame.

Goatman13 commented:
> It's nothing ^^ , diagnosing floats is more tricky than some usual tests, because the wrong calculation can happen way before we even noticed it went wrong.
> 
> In the case of constantine, they really seems to pre-calculate a lot of stuff while loading, I guess to save on compute time while ingame.

Normally I'm aware of that, but here they tricked me with working rounding modes after loading. :D

Ok, next test is Tales of Destiny Directors Cut (SLPS-25842), reason why pcsx2 have one of most ugly per game hacks in FPU recompiler. Game works now as intended in interpreter with accurate Mul/Div. Tested in place where FpuMul hack is required in recompiler.  Attached card if someone need it, use continue option. Be aware that save is using fast save slot, it will be deleted from memory card after loading one time.
[Mcd001.zip](https://github.com/user-attachments/files/17775905/Mcd001.zip)

Goatman13 commented:
Dirge of Cerberus (bug is exclusive to SLPM-66271) https://forums.pcsx2.net/Thread-Bug-Report-Dirge-of-Cerberus-Final-Fantasy-VII-NTSC-J-SLPM-66271

Work with accurate VU0 mul/div. Require disabled EE interpreter but not any of accurate fpu modes. Could be that pcsx2 don't fallback to VU0 int on COP2 opcodes when EE recompiler is selected. Not sure if it is worth looking into at this point.
Memcard: [VMC0.zip](https://github.com/user-attachments/files/17776939/VMC0.zip)
Same like Constantine, game use fast save. After loading one time, it will disappear.

GitHubProUser67 commented:
You are right, VU0 soft-float options also affects the COP2, but to enable accurate COP2 you need to disable the EE Recompiler.

F0bes commented on pcsx2/Ps2Float.cpp:
```diff
@@ -0,0 +1,791 @@
+	else if (rawExp <= 0)
+		return Ps2Float(man < 0, 0, 0);
+
+	return Ps2Float((uint32_t)man & SIGNMASK | (uint32_t)rawExp << 23 | ((uint32_t)absMan & 0x7FFFFF));
```
```suggestion
	return Ps2Float(((uint32_t)man & SIGNMASK) | (uint32_t)rawExp << 23 | ((uint32_t)absMan & 0x7FFFFF));
```
Should address MSVC warning -Wbitwise-op-parentheses

F0bes commented on pcsx2/Ps2Float.cpp:
```diff
@@ -0,0 +1,791 @@
+		return add ? Min() : Ps2Float(0x7F7FFFFE);
+
+	Console.Error("Unhandled abnormal add/sub floating point operation");
+}
```
You don't return anything here despite your function signature. It might be an abnormal case but you still need to make it deterministic (this is currently UB).
Should address MSVC warning -Wreturn-type

F0bes commented on pcsx2/Ps2Float.cpp:
```diff
@@ -0,0 +1,791 @@
+	}
+
+	Console.Error("Unhandled abnormal mul/div floating point operation");
+}
```
You don't return anything here despite your function signature. It might be an abnormal case but you still need to make it deterministic (this is currently UB).
Should address MSVC warning -Wreturn-type

F0bes reviewed (changes requested):
Gave it a quick glance.
Can you use `u32` and `int` (or even better, `s32`) instead of the `int32_t` and `uint32_t` typedefs?
Same for `u8`,`u16` etc.

TheLastRar commented on pcsx2/VUops.cpp:
```diff
@@ -923,498 +916,540 @@ static __fi void _vuMULq(VURegs* VU)
 	VECTOR* dst;
 	if (_Fd_ == 0)
 		dst = &RDzero;
 	else
 		dst = &VU->VF[_Fd_];
 
-	if (_X) dst->i.x = VU_MACx_UPDATE(VU, vuDouble(VU->ACC.i.x) + ( vuDouble(VU->VF[_Fs_].i.x) * vuDouble(VU->VF[_Ft_].i.x))); else VU_MACx_CLEAR(VU);
-    if (_Y) dst->i.y = VU_MACy_UPDATE(VU, vuDouble(VU->ACC.i.y) + ( vuDouble(VU->VF[_Fs_].i.y) * vuDouble(VU->VF[_Ft_].i.y))); else VU_MACy_CLEAR(VU);
-    if (_Z) dst->i.z = VU_MACz_UPDATE(VU, vuDouble(VU->ACC.i.z) + ( vuDouble(VU->VF[_Fs_].i.z) * vuDouble(VU->VF[_Ft_].i.z))); else VU_MACz_CLEAR(VU);
-    if (_W) dst->i.w = VU_MACw_UPDATE(VU, vuDouble(VU->ACC.i.w) + ( vuDouble(VU->VF[_Fs_].i.w) * vuDouble(VU->VF[_Ft_].i.w))); else VU_MACw_CLEAR(VU);
+	tmp = &VU->TMP;
+	if (_X) {tmp->i.x = vuAccurateMulDiv(VU, VU->VF[_Fs_].i.x, VU->VF[_Ft_].i.x, 0); dst->i.x = VU_MACx_UPDATE(VU, vuAccurateAddSub(VU, VU->ACC.i.x, tmp->i.x, 0));} else VU_MACx_CLEAR(VU);
+	if (_Y) {tmp->i.y = vuAccurateMulDiv(VU, VU->VF[_Fs_].i.y, VU->VF[_Ft_].i.y, 0); dst->i.y = VU_MACy_UPDATE(VU, vuAccurateAddSub(VU, VU->ACC.i.y, tmp->i.y, 0));} else VU_MACy_CLEAR(VU);
+	if (_Z) {tmp->i.z = vuAccurateMulDiv(VU, VU->VF[_Fs_].i.z, VU->VF[_Ft_].i.z, 0); dst->i.z = VU_MACz_UPDATE(VU, vuAccurateAddSub(VU, VU->ACC.i.z, tmp->i.z, 0));} else VU_MACz_CLEAR(VU);
+	if (_W) {tmp->i.w = vuAccurateMulDiv(VU, VU->VF[_Fs_].i.w, VU->VF[_Ft_].i.w, 0); dst->i.w = VU_MACw_UPDATE(VU, vuAccurateAddSub(VU, VU->ACC.i.w, tmp->i.w, 0));} else VU_MACw_CLEAR(VU);
```
What's the reason for using a global TMP vector?
This can just be using a locally defined one, right?

ditto for the other fused ops

GitHubProUser67 replied:
The reason behind the global vector is to allow for more generalised debugging when doing an FMA operation.

The use cases where it can be used, is when you don't know where you float went wrong, and it happened in an FMA instruction, so you could simply hook the global TMP and see all the values the FMA processed in between.

Ultimately you decide if it is worth or not.

TheLastRar reviewed (commented):
I had skimmed though this a little while ago and this was what I picked up on.

TellowKrinkle reviewed (commented):
For more general design, I think PS2Float should hold the float directly, with methods to extract the pieces, rather than storing the parts separately.  This would save deconstructing and then reconstructing floats to do operations like abs and negate, which really just need to `and` or `xor` with some bits.

e.g.
```cpp
class PS2Float
{
public:
	u32 raw;
	constexpr explicit PS2Float(u32 raw_): raw(raw_) {}
	constexpr u32 Mantissa() const { return (raw & 0x7fffff) | 0x800000; }
	constexpr u8 Exponent() const { return (raw >> 23) & 0xff; }
	constexpr bool Sign() const { return raw >> 31; }
	constexpr PS2Float Abs() const { return { raw & 0x7fffffff }; }
	constexpr PS2Float Negate() const { return { raw ^ 0x80000000 }; }
};
```

(If we really want to let the compiler optimize well, we should implement add/sub/etc as static functions that take both lhs and rhs by value, then we can mark them with `__attribute__((const))` and let the compiler assume they won't touch any global variables or anything, which might be important if we're hoping the compiler will hoist all those `CHECK_VU_SOFT_ADDSUB` checks outside of the four if statements in the VUs, though I'm still kind of doubtful that it will)

TheLastRar commented on pcsx2-qt/Settings/AdvancedSettingsWidget.ui:
```diff
@@ -33,8 +33,8 @@
         <y>-447</y>
-        <width>790</width>
-        <height>1049</height>
+        <width>793</width>
```
Blame Qt designer for that one, I'm not sure why it put those 3 pixels in

Goatman13 commented:
https://github.com/PCSX2/pcsx2/issues/5070 fixed with accurate VU0 Mul/Div for COP2 (require EE interpreter like Dirge of Cerberus).

Goatman13 commented:
https://github.com/PCSX2/pcsx2/issues/4546 fixed with accurate VU1 Mul/Div. :)

TheLastRar commented on pcsx2/CMakeLists.txt:
```diff
@@ -93,8 +93,7 @@ set(pcsx2Sources
 	Patch.cpp
-	BoothMultiplier.cpp
-	Ps2Float.cpp
+	PS2Float.cpp
```
The files will also need an uppercase S for CMake to be able to find them on Unix systems

TheLastRar replied:
Ps2Float.h (in addition to the now corrected PS2Float.cpp) will also need a capital S in its filename.

GitHubProUser67 commented on pcsx2/VUops.cpp:
```diff
@@ -462,45 +463,82 @@ static __fi float vuDouble(u32 f)
-	if (_Z){ VU->VF[_Ft_].f.z = fabs(vuDouble(VU->VF[_Fs_].i.z)); }
-	if (_W){ VU->VF[_Ft_].f.w = fabs(vuDouble(VU->VF[_Fs_].i.w)); }
+	if (CHECK_VU_SOFT_ADDSUB((VU == &VU1) ? 1 : 0) || CHECK_VU_SOFT_MULDIV((VU == &VU1) ? 1 : 0) || CHECK_VU_SOFT_SQRT((VU == &VU1) ? 1 : 0))
+	{
+		if (_X)
+		{
+			VU->VF[_Ft_].i.x = PS2Float(VU->VF[_Fs_].i.x).Abs();
+		}
+		if (_Y)
+		{
+			VU->VF[_Ft_].i.y = PS2Float(VU->VF[_Fs_].i.y).Abs();
+		}
+		if (_Z)
+		{
+			VU->VF[_Ft_].i.z = PS2Float(VU->VF[_Fs_].i.z).Abs();
+		}
+		if (_W)
+		{
+			VU->VF[_Ft_].i.w = PS2Float(VU->VF[_Fs_].i.w).Abs();
+		}
+	}
+	else
+	{
+		if (_X)
+		{
+			VU->VF[_Ft_].f.x = fabs(vuDouble(VU->VF[_Fs_].i.x));
+		}
+		if (_Y)
+		{
+			VU->VF[_Ft_].f.y = fabs(vuDouble(VU->VF[_Fs_].i.y));
+		}
+		if (_Z)
+		{
+			VU->VF[_Ft_].f.z = fabs(vuDouble(VU->VF[_Fs_].i.z));
+		}
+		if (_W)
+		{
+			VU->VF[_Ft_].f.w = fabs(vuDouble(VU->VF[_Fs_].i.w));
+		}
+	}
```
I will always use the PS2 Float version, the problem is not fabs directly, but the VUDouble that is applied before, it can mess up some floats.

GitHubProUser67 commented:
There are 3 code styles remaining, The removal of the Global VU tpm, PCSX2's own helpers in PS2Float and improvement of the PS2Float class to store the raw float.

Goatman13 commented:
At some point you changed back mantissa normalization constant for Div. From 0x39 to 0x40 which is used in IEEE754. 
I'm just curious if previous value was incorrect and you wanted to use 0x3F initially? I didn't test anything, just noticed that 0x39 is little bit weird here, like confused hex with decimal accidentally.
This commit: https://github.com/PCSX2/pcsx2/pull/12001/commits/b0b65fa2489e681090dca9e72cb365a1819e1573

GitHubProUser67 commented:
I see you are well documented on the subject :) I will explain to you why this "experiment" failed.

The current Div/Sqrt system mimic the PS3 SPEs (I have more experience with that system), and they uses a IEEE "like" divisor that allowed me to test what is closer to PS2 hardware.

The PS2 Div is a mess, they round to nearest by default, but their "rounding" is completely custom. So I took the PS3 as a basis and tried to "lower" the rounding delta to see what is closer to PS2 hardware, 0x39 seemed fine, but in actual fact, that was a hack.

Changing this value introduced a slight BIAS in the result, which can affect high floats values and makes them incorrect.

Current Div/Sqrt system is exactly like a PS3 emulating the PS2 in therms of results, which means way closer (and fixes the VAST majority of games), but still one bit off in some edge cases.

Add/Sub/Mul are PS2 perfect.

GitHubProUser67 commented:
I would say the main issue came from the fact it was widely assumed the PS2 and PS3 shared the same "float" calculation behaviours, while in actual fact, they do not.

The IBM floats, even if not compliant to IEEE are still very similar in action.

The PS2 deviates slightly on some defined calculations. The Multiplication is flawed for instance, due to them using a booth and wallace approach (Method which cost a LOT LESS to manufacture) that was dropping bits in some cases, causing plenty of calculation deviation ( 1 * something NOT ALWAYS equal to something).

Goatman13 commented:
> I would say the main issue came from the fact it was widely assumed the PS2 and PS3 shared the same "float" calculation behaviours, while in actual fact, they do not.

That was the plan as far as I know. Probably someone underestimated how important 1:1 implementation was to keep compatibility. PS3 CELL implemented special Altivec/VMX mode to make PPE core vector floats compatible with SPU, but also with PS2 emulation in mind. Later it turned out that they really miscalculated importance of accurate floating point math in PS2 emulation. To the point that first fixes in ~2008-2010 were strictly per game, like "on this pc, add 0x01 to this reg, and get back execution to recompiler". Soft floats, including accurate DIV, came later, when new people were hired. :) In the end only VU1 is using real SPE, everything else use PPE VMX compatibility mode.

>Current Div/Sqrt system is exactly like a PS3 emulating the PS2 in therms of results, which means way closer (and fixes the VAST majority of games), but still one bit off in some edge cases.
>
>Add/Sub/Mul are PS2 perfect.

And that's indeed a great outcome. Nice work. :)

GitHubProUser67 commented:
Thank you Goatman13! The PCSX2 team also gets massive credits here for their help and support, of course the goal is to be PS2 perfect on this one. There is this one game, Mortal Combat Shaolin Monks which doesn't work fully yet on the Soft-Float div due to the game doing loads of Log2 on top of existing Log2 results, leading to it relying on rounding in booth directions...

TellowKrinkle commented:
BTW the reason macOS builds are failing is they have PCH off, which means you don't get a bunch of headers auto-included.  In particular, you need an `#include "common/Pcsx2Defs.h"` at the top of PS2float.c to get the `u32`, etc defines.

GitHubProUser67 commented:
Now the only code style remaining is the Global VU TMP.

TellowKrinkle commented on common/BitUtils.h:
```diff
@@ -84,6 +103,32 @@ namespace Common
+
+	__fi static s32 GetMostSignificantBitPosition(u32 value)
+	{
+		for (s32 i = 31; i >= 0; i--)
+		{
+			if (((value >> i) & 1) != 0)
+				return i;
+		}
+		return -1;
```
```suggestion
#ifdef _MSC_VER
		unsigned long bit;
		return _BitScanReverse(&bit, value) ? bit : -1;
#else
		return value ? __builtin_clz(value) ^ 31 : -1;
#endif
```
And you can drop the rest of the changes and make BitScanReverse8 and clz just call this (or pick one stop using the rest).
As a side note, not handling the value == 0 case will be a bit cheaper, and in my experience in pretty much all cases, the caller has already checked and is only calling this if they know `value` is nonzero, so the separate check here is just wasted cycles.

GitHubProUser67 replied:
Ok, I will do that when possible and push, I admit I lacked experience with the BitScan part.

GitHubProUser67 commented:
I declare Final Fantasy X fixed on PCSX2!

GT4/Tourist Trophy are an edge case that we can hopefully sort out soon.

AmyRoxwell commented:
Hopefully This could, someday, get implemented into the recompiler. Having this many games fixes (And some outright going from broken to playable) is really nice to see! Keep up the great work you doing!

LoStraniero91 commented:
**Superman Returns**
Needs VU1 - MUL/DIV to fix SPS
Note: the SPS is fixed with the "VU Overflow hack", but this causes buildings to fade in and out near the screen edges when flying at high altitude, the soft-float setting completely solves both issues

LoStraniero91 commented:
> so this PR does not fix the bugs in Colin McRae Rally 3, correct?

Correct. Even tried with EE Cache and full floats, no dice.

seta-san commented:
> > so this PR does not fix the bugs in Colin McRae Rally 3, correct?
> 
> Correct. Even tried with EE Cache and full floats, no dice.

either floating point needs to be perfect or it’s a timing issue

vcx33 commented:
> I declare Final Fantasy X fixed on PCSX2!
> 
> GT4/Tourist Trophy are an edge case that we can hopefully sort out soon.

For FFX, the boss turning invisible bug doesn't happen when using the interpreter, but it doesn't seem to be affected by soft floats, and seems to be a regression, this will have to be re tested on the recompiler.

GitHubProUser67 commented:
And there we have our New-Year present! Fully accurate FPU in PCSX2!

Aside a few things remaining to adjust (flags and newer Div/Sqrt method), we now have a solid basis that replicate the PS2 operations.

The last commit fixes:

- https://github.com/PCSX2/pcsx2/issues/11528
- https://github.com/PCSX2/pcsx2/issues/11507
- https://github.com/PCSX2/pcsx2/issues/4760
- https://github.com/PCSX2/pcsx2/issues/9831
- Gran Turismo 4 license tests
- Mortal Combat Shaolin Monks
- Tokyo Xtreme Racer Zero (Accurate Mul/Div for the EE FPU)
- Any other games using the NegDiv hack or any other FPU rounding mode.

seta-san commented:
Jesus. This is amazing. many of the oldest issues are suddenly just fixed. This is truly God's work you and Tellow and everyone else has done here.

LoStraniero91 commented:
> And there we have our New-Year present! Fully accurate FPU in PCSX2!
> 
> Aside a few things remaining to adjust (flags and newer Div/Sqrt method), we now have a solid basis that replicate the PS2 operations.
> 
> The last commit fixes:
> 
> * [[BUG]: FPU rounding issues in Shadow of the Colossus (SCUS-97472) #11528](https://github.com/PCSX2/pcsx2/issues/11528)
> * [[BUG]: Tourist Trophy: License Tests are broken and not completeable, player is placed out of bounds #11507](https://github.com/PCSX2/pcsx2/issues/11507)
> * [[BUG]: Disney's UP TLB Miss #4760](https://github.com/PCSX2/pcsx2/issues/4760)
> * [[BUG]: Need for Speed Undercover not loading save games properly unless they’re created within PCSX2 #9831](https://github.com/PCSX2/pcsx2/issues/9831)
> * Gran Turismo 4 license tests
> * Mortal Combat Shaolin Monks
> * Tokyo Xtreme Racer Zero (Accurate Mul/Div for the EE FPU)
> * Any other games using the NegDiv hack or any other FPU rounding mode.

Tested Colin McRae Rally 3 with this commit and the AI still goes out of sync and crashes into walls. What I should also add is that the AI keeps accelerating against the wall, then it "reminds" it could reverse as I can see it tries to back out from the wall, but it stays in place and goes nowhere. After almost 3 minutes, the AI "dies", as it probably thought it has finished the race.
Very temperamental little game.
![image](https://github.com/user-attachments/assets/6eee34e8-bfc3-4461-a0ed-66279b3d7366)

Kaibayugi2002 commented:
> > And there we have our New-Year present! Fully accurate FPU in PCSX2!
> > Aside a few things remaining to adjust (flags and newer Div/Sqrt method), we now have a solid basis that replicate the PS2 operations.
> > The last commit fixes:
> > 
> > * [[BUG]: FPU rounding issues in Shadow of the Colossus (SCUS-97472) #11528](https://github.com/PCSX2/pcsx2/issues/11528)
> > * [[BUG]: Tourist Trophy: License Tests are broken and not completeable, player is placed out of bounds #11507](https://github.com/PCSX2/pcsx2/issues/11507)
> > * [[BUG]: Disney's UP TLB Miss #4760](https://github.com/PCSX2/pcsx2/issues/4760)
> > * [[BUG]: Need for Speed Undercover not loading save games properly unless they’re created within PCSX2 #9831](https://github.com/PCSX2/pcsx2/issues/9831)
> > * Gran Turismo 4 license tests
> > * Mortal Combat Shaolin Monks
> > * Tokyo Xtreme Racer Zero (Accurate Mul/Div for the EE FPU)
> > * Any other games using the NegDiv hack or any other FPU rounding mode.
> 
> Tested Colin McRae Rally 3 with this commit and the AI still goes out of sync and crashes into walls. What I should also add is that the AI keeps accelerating against the wall, then it "reminds" it could reverse as I can see it tries to back out from the wall, but it stays in place and goes nowhere. After almost 3 minutes, the AI "dies", as it probably thought it has finished the race. Very temperamental little game. ![image](https://private-user-images.githubusercontent.com/29802222/400170575-6eee34e8-bfc3-4461-a0ed-66279b3d7366.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYwMzU4MDgsIm5iZiI6MTczNjAzNTUwOCwicGF0aCI6Ii8yOTgwMjIyMi80MDAxNzA1NzUtNmVlZTM0ZTgtYmZjMy00NDYxLWEwZWQtNjYyNzliM2Q3MzY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA1VDAwMDUwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZkZDY1YjQxODMwZTAxNDE2YTJhMjZlMjFmMzZmN2M3ZTE5NTY5YzAyMmI5MDQ4YmJiMGVlOTYzOWUwNmU0NGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.St8FuaYo2j1wiuuAzYPQb5xw27rXIU0GD0KXxgX98ok)

I'm assume that only testers can download this at the moment. So unless your a tester, we the general public don't have access yet.

Valtekken commented:
> @LoStraniero91 I'm assuming that only testers can download this at the moment. So unless your a tester, we the general public don't have access yet.

You can download it yourself from the artifacts right below the comments.

kamfretoz commented:
> I'm assuming that only testers can download this at the moment. So unless your a tester, we the general public don't have access yet.

You only need a GitHub account to download the CI artifacts.

And you can download it from [here](https://github.com/PCSX2/pcsx2/actions/runs/12613446344)

NIWDERED-07 commented:
> ¡Y ya tenemos nuestro regalo de Año Nuevo! ¡Un FPU totalmente preciso en PCSX2!
> 
> Aparte de algunas cosas que quedan por ajustar (banderas y el nuevo método Div/Sqrt), ahora tenemos una base sólida que replica las operaciones de PS2.
> 
> La última confirmación corrige:
> 
> * [[ERROR]: Problemas de redondeo de FPU en Shadow of the Colossus (SCUS-97472)  #11528](https://github.com/PCSX2/pcsx2/issues/11528)
> * [[ERROR]: Trofeo Turístico: Las pruebas de licencia están rotas y no se pueden completar, el jugador queda fuera de los límites  #11507](https://github.com/PCSX2/pcsx2/issues/11507)
> * [[ERROR]: Disney's UP TLB Miss  #4760](https://github.com/PCSX2/pcsx2/issues/4760)
> * [[ERROR]: Need for Speed ​​Undercover no carga los juegos guardados correctamente a menos que se creen dentro de PCSX2  #9831](https://github.com/PCSX2/pcsx2/issues/9831)
> * Pruebas de licencia de Gran Turismo 4
> * Monjes Shaolin de combate mortal
> * Tokyo Xtreme Racer Zero (multiplicación y división precisas para EE FPU)
> * ¿Algún otro juego que utilice el hack NegDiv o cualquier otro modo de redondeo de FPU?


Wonderful :O, I look forward to them soon being able to repair driver 3, it makes me want to play it again :D

Tokman5 commented:
Demo replays of Tokyo Xtreme Racer 0 are MUCH closer to the ones running on console than ever!
Unfortunately, they are still not equal (I don't know if this is solely due to FPU accuracy). It means the behavior of non-player cars behave differently, so some replays are still broken. Even so, the last commit is still a huge step.

arienap commented:
I did some testing and PAL Stuntman seems massively improved with EE and VU0 soft-float.
The conspiracy level where the NPC vehicle drives on ice seems to be the best test level, it unfortunately still deviates later during the level.
Did not have the time to test all levels as the Interpreter is not full speed on my Ryzen 5700X3D.
This is awesome, referencing the relevant ticket here for reference:
https://github.com/PCSX2/pcsx2/issues/2990

arienap commented:
I tested Stuntman PAL version again on the new build and the "River race" can now be completed in the emulator.
I can do more testing of other maps and also NTSC version, but from what I could see now it seems like Stuntman is now fully accurate to the PS2.
This is awesome work!!

TheLastRar commented:
Something here is causing issues in the VU1 interpreter for Jak 3 (Pal)

Test with EE rec, VU0 rec, VU1 Int with all Softfloat options disabled (or enabled results unchanged), the intro cutscene doesn't render.
This issue does not occur on master.

NIWDERED-07 commented:
> I tested Stuntman PAL version again on the new build and the "River race" can now be completed in the emulator. I can do more testing of other maps and also NTSC version, but from what I could see now it seems like Stuntman is now fully accurate to the PS2. This is awesome work!!

Would you be so kind as to run all the tests and comment if it works? What you have achieved is great, because if everything goes well, possibly the chase missions of driv3r (driver 3) would already be finished so we will finally be able to enjoy that title as well :D

NIWDERED-07 commented:
> > I tested Stuntman PAL version again on the new build and the "River race" can now be completed in the emulator. I can do more testing of other maps and also NTSC version, but from what I could see now it seems like Stuntman is now fully accurate to the PS2. This is awesome work!!
> 
> This might potentially fix #9681 aswell

Could you try to test again and see if the problem is already fixed? pls

GitHubProUser67 commented:
@TheLastRar Thank you for the bug report, unfortunately, this will not be fixed quickly, I will have to completely re-do my initial work on the interpreters, since the current design of the VUs, and the FPU, doesn't allow for U/O/D/I flags propagation. This will be fixed when the PR will be ready.

crashGG commented:
Hi bro,after this PR is merged, will the next step for recompiler be coming soon?

GitHubProUser67 commented:
We will see, I prefer focusing on the interpreter version for now, until is is good to integrate.

GitHubProUser67 commented:
@TheLastRar I identified the problem and where it came from, unfortunately it is due to a timing issue that got ported from this fork : https://github.com/Goatman13/pcsx2/tree/accurate_int_add_sub

The problem resides is how the VU store the multiplier in the x/w/z/y methods:

My PR is like this:
![image](https://github.com/user-attachments/assets/cd91bcc2-3130-4368-8665-dd20f31d90a9)

But it needs (like master) to store the data before doing anything:
![image](https://github.com/user-attachments/assets/1acd16ff-d3fe-4987-93f8-180559223096)

So this is awesome news, I can fix it even with the current design, however, what a game this Jak3! They optimized the thing to it's absolute limits.

LoStraniero91 commented:
![image](https://github.com/user-attachments/assets/52306285-3c88-4dcf-9f9b-52a08a9b7c11)
The latest update has finally fixed Colin McRae Rally 3's wonky AI. As you can see, the opponent is ahead of me and has finished the race without swerving or crashing even once.

NIWDERED-07 commented:
> We will see, I prefer focusing on the interpreter version for now, until is is good to integrate.

Hi, I have a question, is the possible "soft-float" solution integrated in the most recent version of PCSX2? or do I have to download something and put it in the emulator? I want to try Driv3r or should I wait for a next release? For now I can only say that the graphics get distorted when entering any of the game's configuration options, here is an image below, also don't you know that the text breaks when I finish a mission and enter the "save to profile" option, by the way, thanks to everyone for the work you are doing :D
![image](https://github.com/user-attachments/assets/683ae4f3-a2fb-4fdb-98df-5f166e6036c0)

terremoth commented:
OMG Awesome to see this issue!

I will try playing ASAP Driv3r too if it fixes my issue too #11714

F0bes commented:
> The Pull Request is now officially ready for a code review.

Your macos builds are still failing.

terremoth commented:
It looks like for some reason the MacOS building is not recognizing the `u32` and `s32` types 

https://github.com/PCSX2/pcsx2/actions/runs/13348284205/job/37285797895?pr=12001#step:14:5744

F0bes commented:
MacOS doesn't use precompiled headers. You will need to include the header where those are typedef'd.

GitHubProUser67 commented:
MAC_OS should be fixed now.

seta-san commented:
> MAC_OS should be fixed now.

Codacy has also stopped complaining.

TellowKrinkle reviewed (changes requested):
- I see a lot of `if (CHECK_FPU_SOFT_ADDSUB || CHECK_FPU_SOFT_MULDIV || CHECK_FPU_SOFT_SQRT)`.  You can put all of them in one bitfield and check the whole thing with a single instruction like `if (FPU_SOFTFLOAT_FLAGS & (SOFTFLOAT_ADDSUB | SOFTFLOAT_MULDIV | SOFTFLOAT_SQRT))`.
- You seem to have decided to reimplement a whole bunch of instructions that were already either already accurate or nearly accurate.  Please don't.  PS2Float.h should contain the following operations, which are actually complicated enough to need emulation in a dedicated file: add, mul, div, madd, sqrt.  The rest should be left alone, and made accurate in place if necessary.  Things like ftoi and itof should just be accurate all the time, there's no point in saving the 4 instructions to not handle the edge cases.  Don't worry about accuracy with rounding modes other than chop/zero.  I've opened PRs for accurate ftoi and clip, so assuming being accurate all the time doesn't break any games, let's go with that.
- I don't want a div/sqrt algorithm that loops over individual bits in the divisor/dividend.  If you'd like me to make a sqrt version of the div I posted to discord I can.  Also please move the div code to PS2Float.cpp with the rest of the stuff.
- Please leave the old MAC update function as is and make a separate one for PS2Floats.  #12323 will probably make this a lot easier.

MarioSonic2987 commented:
Deadly Strike needs Soft-Float Multiplication/Division to fix overlapping tiles that can hide character models.
![image](https://github.com/user-attachments/assets/9d2a0a1a-0860-4e9c-a027-bc6eacd74e96)

Soft-Float Multiplication/Division disabled:
![image](https://github.com/user-attachments/assets/d7cdef72-008c-4bc9-b238-11f9405e8a09)

terremoth commented:
I am super excited for this fix

chaoticgd commented:
Another issue to add to your list. I haven't tested it but it seems likely soft float will fix this: https://github.com/PCSX2/pcsx2/issues/12160

AmyRoxwell commented:
> Another issue to add to your list. I haven't tested it but it seems likely soft float will fix this: #12160

Tested this with the current version, it didn't help the models legs disappearing, the game getting softlocked neither the other issues.

TheLastRar commented:
Coming back to retest #12200 
The fight with the Dark Satellite specifically needs EE Soft-Float Multiplication/Division during its tentacle attacks

LoStraniero91 commented:
Pretty much every replay for any racing game (including Stuntman and Driv3r) will require tight FPU accuracy.
So far I've tested different games with a replay saved on PS2, then playing them with a back-to-back PCSX2-PS2 comparison:
**Driv3r** (All floats enabled for VU0)
**Gran Turismo 4** (All floats enabled for EE, or it will even fail to load the replay)
**WRC Rally Evolved** (All floats enabled for EE and VU0)

Dynamo128 commented:
Out of curiosity, to any testers: what is the current state of Stuntman with this PR? Has anyone done thorough testing on both the NTSC and PAL versions? If so, can you point out which missions break?

IzhenPS2 commented:
what should i enable to fix Need for Speed Undercover #9831 bug?
i tried enabling all floats for EE and VU0 and still hasn't found a solution, am i missing something?

GitHubProUser67 commented:
This pull request will be closed in favor of a newer one since the rebasing process cannot be done.

GitHubProUser67 committed:
Core/EE: Implements Soft-Floats for the interpreters.

This commit implements software floating points support for the interpreters in PCSX2's interpreters.nit specification.

This work is a combination or several efforts and researches done prior.

Credits:

- https://www.gregorygaines.com/blog/emulating-ps2-floating-point-nums-ieee-754-diffs-part-1/

- https://github.com/GitHubProUser67/MultiServer3/tree/main/BackendServices/PS2FloatLibrary

- https://github.com/Goatman13/pcsx2/tree/accurate_int_add_sub

- PCSX2 Team for their help and support in this massive journey.

Fixes codacity warnings.

Mrlinkwii commented:
just so we can  link it the new pr is #12550

F0bes commented:
`git reset --hard origin/master`

Creating a new PR causes information to now be split.

TheLastRar commented:
I forgot to mention at the time, but this PR was updated with the changes from https://github.com/PCSX2/pcsx2/pull/12550 and rebased.

Meaning this PR is the updated one.

GovanifY commented:
Hey, I rebased your PR/removed your merge commit to make it apply cleanly against master :)

GitHubProUser67 commented:
What would I do without you? :)

TellowKrinkle reviewed (commented):
In Discord, you claimed this was ready for review, but you haven't addressed any of the following things I pointed out in the last review or written anything detailing why you didn't address them.
- BitScanReverse8 is still there
- I mentioned switching option checks to a bitfield last time, and you haven't done that or given a reason why you don't think that's a good idea.
- I asked you to keep PS2Float simple, and leave out things that the emulator already implements accurately.  Last time, I requested that you keep it to just add, sub, mul, madd, msub, div, and sqrt, but I'll add the transcendentals into that.  But we don't need things like ftoi/itof or comparisons.  We also don't need a ToString.
- I questioned having all the functions take a PS2Float, which now includes a bunch of output-only flags.  Either make PS2Float not contain output flags, or implement the functions as static methods that take `u32`s, with madd/msub taking an extra boolean indicating incoming overflow.  This not only improves performance, but helps make clear what is and isn't needed to people looking at the method signatures.  e.g. someone could easily see that the output of add depends only on the two incoming numbers, but madd actually cares about whether the previous number overflowed or not.

In addition, I posted [this test elf](https://github.com/user-attachments/files/20429047/test.elf.zip) on Discord before the rewrite, you replied that your C# code correctly handles the broken add result, yet it's still broken after you rewrote the C++ code (along with a lot of the flag tests):
<details>
<summary>Test Results</summary>

```
[    1.6582] ADD.S 00000000 + 807FFFFF => 80000000 != 00000000
[    1.6588] VADD 00000000 + 807FFFFF => 80000000 != 00000000
[    1.6591] VADD 00000000 + 807FFFFF MAC FLAGS ZS-- != Z---
[    1.6594] VADD 00800001 + 80800000 MAC FLAGS Z--- != Z-U-
[    1.6597] VADD {3F800000 7FFFFFFF 00800000 00800001} + {BF800000 FFFFFFFF 80800000 80800000} STATUS FLAGS ZS----------Z----- != ZS--US------Z-U---
[    1.6599] VADD 80800001 + 00800000 MAC FLAGS ZS-- != ZSU-
[    1.6603] VADD {BF800000 FFFFFFFF 80800000 80800001} + {3F800000 7FFFFFFF 00800000 00800000} STATUS FLAGS ZSSS--------ZS---- != ZSSSUS------ZSU---
[    1.6605] VADD 00900001 + 80900000 MAC FLAGS Z--- != Z-U-
[    1.6607] VADD 00900000 + 808FFFFF MAC FLAGS Z--- != Z-U-
[    1.6609] VADD 80900001 + 00900000 MAC FLAGS ZS-- != ZSU-
[    1.6611] VADD 80900000 + 008FFFFF MAC FLAGS ZS-- != ZSU-
[    1.6615] VADD {00900001 00900000 80900001 80900000} + {80900000 808FFFFF 00900000 008FFFFF} STATUS FLAGS ZSSS--------ZS---- != ZSSSUS------ZSU---
[    1.6617] SUB.S 00000000 - 007FFFFF => 80000000 != 00000000
[    1.6620] VSUB 00000000 - 007FFFFF => 80000000 != 00000000
[    1.6623] VSUB 00000000 - 007FFFFF MAC FLAGS ZS-- != Z---
[    1.6625] VSUB 00800001 - 00800000 MAC FLAGS Z--- != Z-U-
[    1.6628] VSUB {3F800000 7FFFFFFF 00800000 00800001} - {3F800000 7FFFFFFF 00800000 00800000} STATUS FLAGS ZS----------Z----- != ZS--US------Z-U---
[    1.6630] VSUB 80800001 - 80800000 MAC FLAGS ZS-- != ZSU-
[    1.6633] VSUB {BF800000 FFFFFFFF 80800000 80800001} - {BF800000 FFFFFFFF 80800000 80800000} STATUS FLAGS ZSSS--------ZS---- != ZSSSUS------ZSU---
[    1.6635] VSUB 00900001 - 00900000 MAC FLAGS Z--- != Z-U-
[    1.6638] VSUB 00900000 - 008FFFFF MAC FLAGS Z--- != Z-U-
[    1.6640] VSUB 80900001 - 80900000 MAC FLAGS ZS-- != ZSU-
[    1.6642] VSUB 80900000 - 808FFFFF MAC FLAGS ZS-- != ZSU-
[    1.6645] VSUB {00900001 00900000 80900001 80900000} - {00900000 008FFFFF 80900000 808FFFFF} STATUS FLAGS ZSSS--------ZS---- != ZSSSUS------ZSU---
[    1.6651] VMUL 00800000 * 3F000000 MAC FLAGS Z--- != Z-U-
[    1.6654] VMUL 80800000 * 3F000000 MAC FLAGS ZS-- != ZSU-
[    1.6657] VMUL 00800000 * BF000000 MAC FLAGS ZS-- != ZSU-
[    1.6659] VMUL 80800000 * BF000000 MAC FLAGS Z--- != Z-U-
[    1.6663] VMUL {00800000 80800000 00800000 80800000} * {3F000000 3F000000 BF000000 BF000000} STATUS FLAGS ZSSS--------ZS---- != ZSSSUS------ZSU---
[    1.6665] VMUL 20000000 * 1F800000 MAC FLAGS Z--- != Z-U-
[    1.6667] VMUL A0000000 * 1F800000 MAC FLAGS ZS-- != ZSU-
[    1.6669] VMUL 20000000 * 9F800000 MAC FLAGS ZS-- != ZSU-
[    1.6672] VMUL {20000000 A0000000 20000000 A0400000} * {1F800000 1F800000 9F800000 9FC00000} STATUS FLAGS ZSSS--------ZS---- != ZSSSUS------ZSU---
[    1.6674] VMUL 3F080000 * 00C80000 MAC FLAGS Z--- != Z-U-
[    1.6677] VMUL {408F0000 40CF0000 3F480000 3F080000} * {7E8FFFFF 7ECFFFFF 00C80000 00C80000} STATUS FLAGS ZS----OS----Z--O-- != ZS--USOS----Z-UO--
[    1.6680] MADD.S 007FFFFF + 80800000 * 80800000 FLAGS ------------ != SU----------
[    1.6682] MADD.S BF800000 + 80800000 * 80800000 FLAGS ------------ != SU----------
[    1.6686] VMADD {00000000 00000000 00000000 00000000} + {00000000 00000000 80000000 80000000} * {00000000 80000000 00000000 80000000} STATUS FLAGS ZS----------Z----- != ZSSS--------Z-----
[    1.6689] VMADD {00800000 80000000 80000000 007FFFFF} + {80000000 80800000 80000000 80800000} * {80000000 80000000 80800000 80800000} STATUS FLAGS ZS----------Z----- != ZS--US------Z-----
[    1.6693] VMADD {00800000 3F800000 BF800000 BF800000} + {80000000 80800000 80000000 80800000} * {80000000 80000000 80800000 80800000} STATUS FLAGS --SS---------S---- != ZSSSUS-------S----
[    1.6695] VMADD 00900001 + A0400000 * 1FC00000 MAC FLAGS Z--- != Z-U-
[    1.6697] VMADD 808FFFFF + A0400000 * 9FC00000 MAC FLAGS Z--- != Z-U-
[    1.6700] VMADD 80900000 + A0400000 * 9FC00002 MAC FLAGS Z--- != Z-U-
[    1.6703] VMADD {00900000 00900001 808FFFFF 80900000} + {A0400000 A0400000 A0400000 A0400000} * {1FC00000 1FC00000 9FC00000 9FC00002} STATUS FLAGS ZS----------Z----- != ZSSSUS------Z-U---
[    1.6705] VMADD 80900001 + A0400000 * 9FC00000 MAC FLAGS ZS-- != ZSU-
[    1.6707] VMADD 008FFFFF + A0400000 * 1FC00000 MAC FLAGS ZS-- != ZSU-
[    1.6710] VMADD 00900000 + A0400000 * 1FC00002 MAC FLAGS ZS-- != ZSU-
[    1.6713] VMADD {80900000 80900001 008FFFFF 00900000} + {A0400000 A0400000 A0400000 A0400000} * {9FC00000 9FC00000 1FC00000 1FC00002} STATUS FLAGS ZSSS--------ZS---- != ZSSSUS------ZSU---
[    1.6715] MSUB.S 007FFFFF - 80800000 * 00800000 FLAGS ------------ != SU----------
[    1.6717] MSUB.S BF800000 - 80800000 * 00800000 FLAGS ------------ != SU----------
[    1.6725] VMSUB {00000000 00000000 00000000 00000000} - {00000000 00000000 80000000 80000000} * {80000000 00000000 80000000 00000000} STATUS FLAGS ZS----------Z----- != ZSSS--------Z-----
[    1.6729] VMSUB {807FFFFF 80000000 80000000 007FFFFF} - {80000000 807FFFFF 80000000 807FFFFF} * {00000000 00000000 007FFFFF 007FFFFF} STATUS FLAGS ZS----------Z----- != ZSSS--------Z-----
[    1.6732] VMSUB {00800000 80000000 80000000 007FFFFF} - {80000000 80800000 80000000 80800000} * {00000000 00000000 00800000 00800000} STATUS FLAGS ZS----------Z----- != ZSSSUS------Z-----
[    1.6736] VMSUB {00800000 3F800000 BF800000 BF800000} - {80000000 80800000 80000000 80800000} * {00000000 00000000 00800000 00800000} STATUS FLAGS --SS---------S---- != ZSSSUS-------S----
[    1.6738] VMSUB 00900001 - A0400000 * 9FC00000 MAC FLAGS Z--- != Z-U-
[    1.6740] VMSUB 808FFFFF - A0400000 * 1FC00000 MAC FLAGS Z--- != Z-U-
[    1.6742] VMSUB 80900000 - A0400000 * 1FC00002 MAC FLAGS Z--- != Z-U-
[    1.6745] VMSUB {00900000 00900001 808FFFFF 80900000} - {A0400000 A0400000 A0400000 A0400000} * {9FC00000 9FC00000 1FC00000 1FC00002} STATUS FLAGS ZS----------Z----- != ZSSSUS------Z-U---
[    1.6747] VMSUB 80900001 - A0400000 * 1FC00000 MAC FLAGS ZS-- != ZSU-
[    1.6749] VMSUB 008FFFFF - A0400000 * 9FC00000 MAC FLAGS ZS-- != ZSU-
[    1.6751] VMSUB 00900000 - A0400000 * 9FC00002 MAC FLAGS ZS-- != ZSU-
[    1.6754] VMSUB {80900000 80900001 008FFFFF 00900000} - {A0400000 A0400000 A0400000 A0400000} * {1FC00000 1FC00000 9FC00000 9FC00002} STATUS FLAGS ZSSS--------ZS---- != ZSSSUS------ZSU---
[    1.6757] VMSUB {FFFFFFFF 00000000 FFFFFFFF FFFFFFFF} - {FFFFFFFF 7F800000 7FFFFFFF 7FFFFFFF} * {7FFFFFFF FF800000 BF800001 BF800000} STATUS FLAGS ZS----OS----Z--O-- != ZSSS--OS----Z--O--
[    1.6759] VMSUB {3F800000 00000000 73800000 74000000} - {7FFFFFFF 7FFFFFFF 7FFFFFFF 7FFFFFFF} * {BF800000 BF800000 BF800000 BF800000} STATUS FLAGS ------OS-------O-- != --SS--OS-------O--
```
</details>

GoodMornang commented:
how do you download this and put into pcsx2?

seta-san commented:
> how do you download this and put into pcsx2?

This isn’t complete nor is it tested. If you were inclined to want to test it you need to go to the “checks” part of this pr. Choose your OS and then download from the artifacts.

GoodMornang commented:
> > how do you download this and put into pcsx2?
> 
> This isn’t complete nor is it tested. If you were inclined to want to test it you need to go to the “checks” part of this pr. Choose your OS and then download from the artifacts.

Ok. I'm just not very familiar with GitHub but thanks anyway.

kunit1 commented:
What's the plan for this PR?

terremoth commented:
Wooow, updates! 👏👏🚀

Maxkatsur commented:
Any new updates this March? It isn't ready yet.

pstef commented:
1. What should `DoAdd` return when the normalized exponent underflows?

On my 90K, the adder does not produce +/-0 on underflow. For example:

| a | b | op | hardware result | `PS2Float` result |
|---|---|----|-----------------|-------------------|
| `0x0080008B` | `0x008000EA` | sub.s | **`0x803E0000`** (U\|SU set) | `0x80000000` |
| `0x8080002C` | `0x008000EC` | add.s | **`0x00400000`** (U\|SU set) | `0x00000000` |
| `0x80800036` | `0x00829F6A` | add.s | **`0x0027CD00`** (U\|SU set) | `0x00000000` |
| `0x80C1AA65` | `0x01200000` | add.s | **`0x007CAB36`** (U\|SU set) | `0x00000000` |

These look like exponent-0 encodings of the fully normalized mantissa: take the cancellation result, normalize it, then store exp=0 with the fraction bits as-is (implicit bit shifted out). In `DoAdd`, at the `rawExp < 1` branch, `absMan` already holds exactly that normalized 24-bit mantissa. Is returning `PS2Float(man < 0, 0, 0)` there intentional, or should it keep `absMan & 0x7FFFFF`? The degenerate case (`0x00800001 - 0x00800000` -> mantissa 1, shift 23 -> fraction 0 -> `+0`) is the one my directed tests tend to hit, which may be why this never showed up. 161/65,536 of my biased vectors hit nonzero remnants, and the same values reproduce on VU0 `vadd.x`/`vsub.x`, so it seems to be a shared adder design.

2. Should `DIV_S`/`SQRT_S`/`RSQRT_S` ever set O/U in FCR31?

For `div.s 0x00800000 / 0x40000000` (underflow) my 90K returns `+0` with FCR31 == `0x01000001`: no U or SU. For `div.s 0x5AE20690 / 0x00800044` (overflow) it returns `0x7FFFFFFF` with FCR31 untouched: no O or SO. The same for sqrt/rsqrt. The soft path forwards the model's `of`/`uf` via `checkOverflowUnderflowSoft` in `DIV_S`/`SQRT_S`/`RSQRT_S`. Is there a revision where the div unit actually raises O/U or should those calls only exist for the FMAC-pipeline ops?

3. When does `rsqrt.s` raise the I flag?

For `rsqrt.s fd, 0x3F80FD6C, 0xBF83566A` (negative ft) I read I|SI on hardware, but `PS2Float::Rsqrt` gives `iv == false`: it strips the sign before calling `Sqrt()`, so the negative-operand path can't trigger. For `rsqrt.s fd, 0xFF3CF542, 0x80000000` (ft = -0) hardware gives `0xFFFFFFFF` with both D|SD and I|SI. Same on VU0 `vrsqrt`. Is the sign-strip in `Rsqrt` deliberate, and if so where should the I flag for negative ft come from?

4. How does the FPU interpreter ever take the `acc_oflw` path in `DetermineMacException`?

After `adda.s Fmax,Fmax` then `madd.s fd, 0x7FFFFFFF, 0xBF800000`, my 90K returns `0x7FFFFFFF` with O set: the latched-ACC behavior the model implements. But the soft interpreter gives `0x00000000`: `fpuAccurateMulAdd()` constructs `PS2Float(_FAValUl_)` fresh, so `of` is always false and the `acc_oflw` branch looks unreachable from `FPU.cpp`. Is the plan to persist the ACC overflow bit somewhere (the x86-full path already keeps `fpuRegs.ACCflag`)? FWIW, once I feed the latch in on my local copy, my entire 65,536-vector FMAC corpus - including `mula`-latching and forced collision cases - matches hardware 100%, so the `DetermineMacException` logic itself appears to be exactly right.

5. Should a zero result clear the U MAC flag on VU?

`vmul.x` of `0x16A08BD5 * 0x03F4FE21` underflows to `+0`; my 90K reports MAC = Z+U together (and Z|U in STATUS). `VU_MAC_UPDATE`'s zero branch masks with `~(0x1100 << shift)`, which clears the U that the underflow branch just set - should that mask be `~(0x1000 << shift)` (clear only O)? For the nonzero underflow remnants from question 1, VU still sets Z+U: Z seems to reflect the flush condition rather than the literal zero-ness of the result, which the existing `0x0101` underflow branch already implies.

6. Where should the product stage's underflow go in `MulAdd`?

`adda.s 0x00800000, 0x00800000` then `madd.s fd, 0x00800000, 0x3F000000` (product flushes, sum doesn't underflow): hardware gives result `0x01000000` with FCR31 = `0x01000009`: SU sticky only, no U cause bit. `PS2Float::MulAdd` ignores `mulres.uf` entirely, so the soft path reports nothing. Across my FMAC corpus, the composition rule that matches hardware everywhere is: cause bits reflect the final result, sticky bits accumulate all internal events including the product flush. Should `MulAdd`/`MulSub` export the product's `uf` as a sticky-only event?

7. How do I reproduce the "PS2 supports denormals in the Mul unit" claim?

I tried in both directions and always got flushes, which is what `Mul()` does in this PR:

- denormal inputs: `mul.s 0x00400000 * 0x71800000 -> +0`, `0x00000001 * 0x7F000000 -> +0`, `0x00400000 * 0x00400000 -> +0` (all flag-silent);
- denormal outputs: `0x00800000 * 0x3F000000` (exact result 2^-127) -> `+0` with U|SU, and no underflowing mul produced a nonzero denormal result, on FPU or VU0.

Was the THPS4 observation perhaps about something else? If there's a specific op sequence, I can run it on hardware directly.

Misc obervations:

- FCR31 reads back `0x01000001` after writing 0: FS (bit 24) and RM=chop (bits 1:0) are hardwired readback bits.
- VU0 status persistence: MAC register and the D/I cause bits survive across ops (`ctc2` to `$vi16` clears stickies only); ZSUO cause bits always reflect the last FMAC op; each div-unit op replaces D/I. VDIV semantics confirmed as modeled: 0/0 -> I (0x10), x/0 and denormal divisors -> D (0x20), sign-XOR saturation to +/-Fmax.
- FPU vs VU multiplier flag parity: both are flag-silent for flushed denormal inputs; both raise U only for true (normal x normal) underflow.

GitHubProUser67 commented:
Ok so, sorry for the recent lack of follow-up, I was taken with exams.

I will respond to each topics:

1: This was effectively an issue in my implementation, the real console does exactly what the @DiscoStarslayer fork do, which also fixed the multimedia subset.

2: This is a fun one, actually... YES, but only in early revisions of the console! For some reasons, they decided to strip these, I don't know yet which unit saw this change.

3: That bit, I was unsure. But when I did the Rsqrt, I assumed it stripped the sign bits, which turns out is flaged as invalid like you said.

4: I had a doubt if the FPU had the same accumulator as the VUs, now confirmed.

5: That's effectively a point I had with the team, that specific flag condition was ported from the original logic, but it was incorrect.

6: I don't know, now that you exposed a test case I can research that aspect.

7: This was actually very wrong, it was caused by other issues in the past code.

Thanks a lot for your report too, will be of a great help.

socialenigma-og commented:
> 2: This is a fun one, actually... YES, but only in early revisions of the console! For some reasons, they decided to strip these, I don't know yet which unit saw this change.

Could it have been related to this?

SCPH-750XX (2005-2006)

"Sony removed the PS1 MIPS R3000A which acted as the I/O CPU, replacing it with a PowerPC CPU. It was in charge of emulating the behavior of the PS1 CPU, and also took control over the network features of the SPEED chip which was the removed from the design of the console. This allowed Sony to finally and completely ditch the hard drive support.

The replacement of the PS1 CPU affected both PS1 and PS2 software. Compatibility fixes were introduced for some of those affected games."

pstef commented on pcsx2/PS2Float.cpp:
```diff
@@ -0,0 +1,733 @@
+		result.iv = true;
+		return result;
+	}
+	u32 am = mantissa(a) << 2;
```
For any operand pair there are only two possible results: T, the truncated 24-bit quotient significand, and T+1. Which of the two hardware returns is a function of the operands, and on a large, precisely defined subset of them it is T, determined before any quotient bit is computed. The 24 selection steps are unnecessary there.

Take the two significands with the hidden bit restored, `sma` and `smb`, and define:

- `lt = 1` when `sma < smb`, else 0; the branch that decides whether the quotient needs the extra shift
- `num = sma << (23 + lt)`
- T = num / smb
- `rem = num - T*smb`, so `0 <= rem < smb`
- `u = smb - rem`, the distance from the exact quotient up to T+1

Then `cap = (sma < smb) ? max(2^23, smb - 2^22) : 2^22`.

My 90K returns T whenever u > cap, with no exceptions across 199,794,412 operand pairs, 150,994,944 of them exhaustive: eighteen divisors, every one of the 2^23 numerator significands at each. 57,612,965 of those pairs satisfy u > cap.

So Div can obtain T from a single 64/32 integer division, test u against cap, and return T when the test succeeds, entering the loop only for the remaining operands. The value returned is bit for bit the one the loop computes on those operands; nothing is approximated.

The same statement holds for Sqrt with cap = 2^23, where X is the radicand placed so its root is a 24-bit integer, R = floor(sqrt(X)) and u = (R+1)^2 - X: no exceptions across all 16,777,216 operands. The trade is worse there, since obtaining R costs an integer square root rather than a division.

pstef commented on pcsx2/PS2Float.cpp:
```diff
@@ -0,0 +1,733 @@
+
+static u64 MulMantissa(u32 a, u32 b)
+{
+	u64 full = static_cast<u64>(a) * static_cast<u64>(b);
```
`if ((full & 0x7FFFFFu) >= 0x8000u)  return full`

RoidyJunkie commented:
Sorry to ask, but I don't take it that most of the soft-float issues are kinda fixed now, are they?

GitHubProUser67 commented:
So, for now I haven't ported yet the fixes to this PR. Will do fairly soon and the PR should be ready with this.

RoidyJunkie commented:
No worries, take the time needed

wxvu commented:
EE Add/Sub + Mul fixes #13871 
<img width="1286" height="955" alt="image" src="https://github.com/user-attachments/assets/440f54f6-f97e-444f-b07f-5bafc6091cdc" />

Maxkatsur commented:
Looking good... may it fix timing, like TimeSplitters 1 Demo issue (https://github.com/PCSX2/pcsx2/issues/14563)?

seta-san commented:
> Looking good... may it fix timing, like TimeSplitters 1 Demo issue (#14563)?

No. This doesn’t fix timing. This makes the math correct.
